lib/Dialect/TTIR/Transforms/TTIRFusing.cpp

```diff
@@ -1874,9 +1879,169 @@ class SplitQueryKeyValueAndSplitHeadsUpdatePattern
+
+    std::size_t dimToConcatWeights = keyMatmulOp.getTransposeB() ? 0 : 1;
+
+    // Assert that keyMatmulOp A and B have rank 2.
+    TT_assertv((keyMatmulOp.getA().getType().getRank() == 2 &&
+                keyMatmulOp.getB().getType().getRank() == 2),
+               "Expected rank 2 for MatMulOp operands A and B");
 
+    ttir::ConcatOp concatenatedWeightMatrix = concatenateAlongDim(
+        rewriter, valueWeightMatrix.getLoc(),
+        ValueRange{keyWeightMatrix, valueWeightMatrix}, dimToConcatWeights);
+
+    Value concatenatedBias = nullptr;
+    if constexpr (std::is_same_v<LinearOp, MatMulOpType>) {
+      Value keyBias = matmulOps[1].getBias();
+      Value valueBias = matmulOps[2].getBias();
+      std::size_t concatDim = matmulOps[1].getBias().getType().getRank() - 1;
+      concatenatedBias =
+          concatenateAlongDim(rewriter, valueBias.getLoc(),
+                              ValueRange{keyBias, valueBias}, concatDim);
+    }
+
+    // Get the original output shape from one of the original linear ops.
+    ArrayRef<int64_t> originalLinearOutputShape =
+        keyMatmulOp.getType().getShape();
+    llvm::SmallVector<int64_t> linearOutputShape(
+        originalLinearOutputShape.begin(), originalLinearOutputShape.end());
+
+    // Multiply the last dimension by 2 (since we're concatenating K, V).
+    linearOutputShape.back() = originalLinearOutputShape.back() * 2;
+
+    // Create linear operation with concatenated weights and bias.
+    RankedTensorType linearOutputType = RankedTensorType::get(
+        linearOutputShape, keyMatmulOp.getType().getElementType());
+
+    Value matmulDpsOutput = rewriter.create<EmptyOp>(
+        keyMatmulOp.getLoc(), linearOutputType.getShape(),
+        linearOutputType.getElementType(), linearOutputType.getEncoding());
+
+    SmallVector<Value> inputs;
+    if (concatenatedBias) {
+      inputs = {reshapeOutput, concatenatedWeightMatrix, concatenatedBias,
+                matmulDpsOutput};
+    } else {
+      inputs = {reshapeOutput, concatenatedWeightMatrix, matmulDpsOutput};
+    }
+
+    MatMulOpType matrixMultOp = rewriter.create<MatMulOpType>(
+        keyMatmulOp.getLoc(), TypeRange{linearOutputType}, inputs,
+        keyMatmulOp->getAttrs());
+
+    TT_assertv(matrixMultOp, "Expected valid matrix multiplication operation");
+
+    // Reshape query matmul / linear op from [batch_size * sequence_size,
+    // query_hidden_size] to [batch_size, sequence_size, query_hidden_size].
+    // Reshape kv matmul / linear op from [batch_size * sequence_size,
+    // 2 * kv_hidden_size] to [batch_size, sequence_size, 2 * kv_hidden_size].
+
+    ArrayRef<int64_t> inputShape = reshapeOp.getInput().getType().getShape();
+    int32_t batchSize = inputShape[I_BATCH_SIZE];
+    int32_t sequenceLength = inputShape[I_SEQUENCE_LENGTH];
+
+    auto queryType = permuteOps[0].getOutput().getType();
+    auto keyType = permuteOps[1].getOutput().getType();
```

Comment:
:warning: **clang\-diagnostic\-error** :warning:
no member named `` getOutput `` in `` mlir::tt::ttir::PermuteOp ``; did you mean `` getInput ``?
```suggestion
    auto keyType = permuteOps[1].getInput().getType();
```